File: src/main/java/xyz/realmath/nilpotent/Field.java
package xyz.realmath.nilpotent;

public interface Field<F extends Field<F>> {

  default boolean isZero() {
    return this.equals(add(thisT()));
  }

  default boolean isOne() {
    return this.equals(multiply(thisT()));
  }

  F add(F f);

  F negate();

  default F minus(F f) {
    return this.add(f.negate());
  }

  F multiply(F f);

  F reciprocal();

  default F dividedBy(F f) {
    return this.multiply(f.reciprocal());
  }

  @SuppressWarnings("unchecked")
  default F thisT() {
    return (F) this;
  }
}

File: src/main/kotlin/xyz/realmath/Field.kt
package xyz.realmath

interface Field<T : Field<T>> : Algebra<T, T> {

    override operator fun times(multiplier: T): T

    fun reciprocal(): T
}

operator fun <T : Field<T>> T.div(f: T): T {
    check(!f.isZero) { "f is zero" }
    return this * f.reciprocal()
}

val <T : Field<T>> T.isOne: Boolean
    get() = this == this * this
